### PR TITLE
[DO NOT REVIEW] AI gateway rate limits design & quick POC

### PR DESCRIPTION
**Pull request by harupy**
[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/9939?quickstart=1)

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

```
mlflow gateway start --config-path examples/gateway/openai/config.yaml
```

```
curl -X 'POST' 'http://127.0.0.1:5000/gateway/chat/invocations' -H 'accept: application/json' -H 'Content-Type: application/json' -d '{
  "max_tokens": 10,
  "messages": [
    {
      "role": "user",
      "content": "hello"
    }
  ]
}'
```

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/gateway/openai/config.yaml [hunk 1]
```diff
@@ -1,6 +1,9 @@
 routes:
   - name: chat
     route_type: llm/v1/chat
+    limit:
+      renewal_period: "minute"
+      calls: 1
```
Comment (harupy) [inline examples/gateway/openai/config.yaml:6]: Global limit vs. per-route limit.

File: mlflow/gateway/app.py [hunk 2]
```diff
@@ -6,8 +6,14 @@
 from fastapi.openapi.docs import get_swagger_ui_html
 from fastapi.responses import FileResponse, RedirectResponse
 from pydantic import BaseModel
+from slowapi import Limiter, _rate_limit_exceeded_handler
```
Comment (harupy) [inline mlflow/gateway/app.py:9]: https://github.com/laurentS/slowapi

File: examples/gateway/openai/config.yaml [hunk 3]
```diff
@@ -1,6 +1,9 @@
 routes:
   - name: chat
     route_type: llm/v1/chat
+    limit:
```
Comment (harupy) [inline examples/gateway/openai/config.yaml:4]: Do we need multiple limit configs?

File: examples/gateway/openai/config.yaml (hunk 1 above)
Comment (serena-ruan) [reply examples/gateway/openai/config.yaml:6]: per-route limit

**Comment by harupy** [comment]
@TomeHirata Would you be interested in working on this?

**Comment by TomeHirata** [comment]
@harupy Yes, please assign this task to me!

**Comment by TomeHirata** [comment]
@harupy I read the doc and understood the overall direction. Before starting the implementation, can I ask if there is a way for me to see [this link](https://github.com/databricks/docs/blob/1a11cec796706750db6538013af930895f8071a4/source/machine-learning/ai-gateway/ai-gateway-tutorial.md)?

**Comment by harupy** [comment]
> can I ask if there is a way for me to see [this link](https://github.com/databricks/docs/blob/1a11cec796706750db6538013af930895f8071a4/source/machine-learning/ai-gateway/ai-gateway-tutorial.md)?

Sorry, this is a link to our internal repo. 

> I also saw mlflow.gateway.get_limits is deprecated. Should I reuse the method? or recreate a new one?

`mlflow.gateway` has been deprecated and replaced by `mlflow.deployments`.

**Comment by TomeHirata** [comment]
Thank you for the reply. So what did the link refer to? Is it the same as the example code in the design doc?
Regarding the scope of the change, can I understand the necessary changes are as follows? Anything I'm missing?

1. Read the rate limit configuration from the endpoint yaml
2. Implement the rate limiting with slowapi
3. Return the rate limit info in the response of get_endpoint

**Comment by harupy** [comment]
> Is it the same as the example code in the design doc?

Yes :)

> Regarding the scope of the change, can I understand the necessary changes are as follows? Anything I'm missing?

Yes, that's correct!

**Comment by TomeHirata** [comment]
Thank you, let me start working on this shortly. Should I take over this PR? or create a new one?

**Pull request closed without merging**